src/nuclide.cpp

```diff
@@ -90,9 +90,21 @@ Nuclide::Nuclide(hid_t group, const std::vector<double>& temperature)
   double T_min = n > 0 ? settings::temperature_range[0] : 0.0;
   double T_max = n > 0 ? settings::temperature_range[1] : INFTY;
   if (T_max > 0.0) {
-    for (auto T : temps_available) {
-      if (T_min <= T && T <= T_max) {
-        temps_to_read.push_back(std::round(T));
+    // For each interval (T_j, T_j+1), if either T_j or T_j+1 are within the
+    // temperature range, load data for *both* T_j and T_j+1
+    int n = temps_available.size();
+    for (int j = 0; j < n; ++j) {
+      if ((T_min <= temps_available[j] && temps_available[j] < T_max)) {
```

Comment:
There is a simpler way to solve this. Just get the index of the minimum temperature to load using `std::lower_bound`, and the index of the maximum temperature to load using `std::upper_bound`. Then load all temperatures spanned by those indices, doing the additional logic to perhaps grab the 0K XS as before. I'm pretty sure this should hit all the edge cases such as having a single temperature matching the data library's temperature perfectly. Another benefit is that this modified version is O(ln(n)) time and the current method is O(n) w.r.t the search but not writing to the `temps_to_read` vector. Not that this piece of the code matters to performance though.

So all this new stuff should be replaceable by three lines of code:

```
...
auto T_min_it = std::lower_bound(temps_available.begin(), temps_available.end(), T_min);
auto T_max_it = std::upper_bound(temps_available.begin(), temps_available.end(), T_max);
std::vector<int> temps_to_read(T_min_it, T_max_it);
```

Also, I'm kind of thinking we should maybe just get rid of "nearest" mode temperature interpolation. I don't really see any performance reason to use this if we can do stochastic interpolation, and users likely expect linear interpolation by default. The benefit here being that we could remove a fair amount of code handling logic for "nearest" mode. Maybe there is some reason to use "nearest" that I'm missing though.

Reply:
Sorry, just one more comment on Gavin's thoughts:
> I would rather have assert(T_max > 0.0) above this. Or gsl's Expects, which is the same thing as far as I can tell.

`assert` is actually ifdef'ed out when the `NDEBUG` preprocessor macro is defined (see [here](https://www.cplusplus.com/reference/cassert/assert/)).  Likewise, CMake's "Debug" configuration adds `-DNDEBUG` to the compiler line (not sure if OpenMC disables that).  So `assert` isn't appropriate for something that is essential to the correct functioning of the code.

Reply:
Nearest temperature mode is needed when you only have a single temperature available and thus can't interpolate

Reply:
Hm, I think we could write the interpolation code to work as expected if only one temperature is available. The only problem would be that this code would use a random number where nearest mode previously hadn't.

```
      for (i_temp = 0; i_temp < kTs_.size() - 1; ++i_temp) {
        if (kTs_[i_temp] <= kT && kT < kTs_[i_temp + 1]) break;
      }

      // Randomly sample between temperature i and i+1
      f = (kT - kTs_[i_temp]) / (kTs_[i_temp + 1] - kTs_[i_temp]);
      if (f > prn(p.current_seed())) ++i_temp;
      break;
    }
```

to

```
      for (i_temp = 0; i_temp < kTs_.size() - 1; ++i_temp) {
        if (kTs_[i_temp] <= kT && kT < kTs_[i_temp + 1]) break;
      }
      i_temp_upper = i_temp+static_cast<int>(kTs_.size() > 1);

      // Randomly sample between temperature i and i+1
      if (kT - kTs_[i_temp] > (kTs_[i_temp_upper] - kTs_[i_temp]) * prn(p.current_seed())) i_temp = i_temp_upper;
      break;
    }
```

Sorry if this is off topic as well. I need to try to keep stuff on topic here...

Reply:
Yes, please try to stay on topic. These are good discussions, but they are easily lost now because they are going to be buried in an unrelated PR. Again, I'd encourage you to open a separate issue or a topic on our discourse forum.